mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala

```diff
@@ -197,8 +229,92 @@ class LocalLDAModel private[clustering] (
   // TODO:
   // override def topicDistributions(documents: RDD[(Long, Vector)]): RDD[(Long, Vector)] = ???
 
+  /**
+   * Calculate and return log variational bound on perplexity using the provided `documents` of
+   * documents as a test corpus.
+   * Perplexity on a test corpus is calculated as:
+   *    perplexity(documents) = exp( -log p(documents) / numWords )
+   * where p is the LDA model. It is upper bounded by the variational distribution using:
+   *    perplexity(documents) <= exp( -(E_q[log p(documents] - E_q[log q(documents)]) / numWords }
+   */
+  def logPerplexity(documents: RDD[(Long, Vector)]): Double = {
+    val numDocs = documents.count()
+    val corpusWords = documents
+      .map { case (_, termCounts) => termCounts.toArray.sum }
+      .sum()
+    val subsampleRatio = numDocs.toDouble / documents.count()
+    val batchVariationalBound = bound(documents, subsampleRatio, docConcentration,
+      topicConcentration, topicsMatrix.toBreeze.toDenseMatrix, gammaShape, k, vocabSize)
+    val perWordBound = batchVariationalBound / (subsampleRatio * corpusWords)
+
+    perWordBound
+  }
+
+
+  /**
+   * Estimate the variational bound of documents from `documents`:
+   *    log p(documents) >= E_q[log p(documents)] - E_q[log q(documents)])
```

Comment:
Oh, sorry, I meant you can remove the equation.  I like the list of params.  But you can do it in a follow-up PR.